metabase/metabase
# Manage column formatting for datasets + add extra form fields (UI)

Adds a few missing dataset field metadata form fields:

* "Special Type" picker (a.k.a field semantic type)
* "Column Formatting" section
* mapped database column field for native datasets (people can select to which real DB column a dataset field is mapped to (this can unlock column-level permission management, sandboxing, etc. on datasets)

⚠️  "Special Type" and "Mapped DB column" pickers need a BE API available to function correctly, so for now there is only a UI for them. Editing these fields won't be persisted or change something on the UI. On the other hand, the "Column Formatting" relies on the card's `visualization_settings`, so it should work properly and you should see immediate feedback when changing formatting options.

### To Verify

1. Open a GUI dataset (they usually have more rich metadata, so you'd see more different states on the UI)
2. Click the dataset name to open the details sidebar and click the "Customize metadata" button
3. Click through a few columns. In the sidebar, you should now see a column formatting section. E.g. for numeric columns, you should see "show progress mini-bar", "number separator" options, etc.
4. Try editing a few "column formatting" options. You should see immediate changes in the results table. Click "Save changes", refresh the page. Changes to column formatting should be persisted
5. Click the "Special type" picker in the field metadata sidebar. It should now expand instead of being a dropdown list. You should also be able to search for a particularly special type by name. ⚠️ This field requires the BE API available and changing it won't currently do anything and the value won't be saved on the server.
6. For a few fields, try selecting a Currency special type (e.g. Cost). A currency picker should appear right behind the special type picker
7. For a few fields, try selecting an FK special type. An FK target picker should appear right behind the special type picker (you should be able to pick to which table an FK corresponds to)
8. Open a native dataset picker. For every field, you should see a "Database column this maps to" field. Click on the select, you should see a picker with a list of database tables. Click on a table, you should see a list of table columns. Click any column, the select should display "TABLE_NAME > FIELD_NAME". ⚠️ This field requires the BE API available and changing it won't currently do anything and the value won't be saved on the server.

### Demo

<details>
<summary>UI + Column Formatting Changes</summary>

https://user-images.githubusercontent.com/17258145/145438909-b0a77fbb-fa37-454c-b6b3-f27b6903be77.mov
</details>

<details>
<summary>Native datasets' "Mapped DB column" field UI demo </summary>

https://user-images.githubusercontent.com/17258145/145439096-997ccd7e-27d2-4f25-a327-3ae2645b1ba7.mov

</details>
